File: 106_asio/Server/Engine/engine/net/protocol/csgProtocol.cpp
#include "engine/net/protocol/csgProtocol.h"
#include "framework/util/endian.h"
#include "../../rpc/rmidef.h"
#include "../../mq/msgBlock.h"
#include "../../rpc/rmiObjectAdapter.h"
#include "../../core/csgIoMgr.h"

int csg::CCsgProtocol::handleRecvData(const CSessionPtr session,const void* inData, const int len)
{
	CAutoLock l(_readLock);
	_recvBuffer->append(inData, len);
#ifdef _DEBUG
	std::string str(_recvBuffer->getData(), _recvBuffer->getDataSize());
	LogDebug("receive is " << str);
#endif

	CCsgIoMgr::instance()->getLogicServer()->post(boost::bind(&CCsgProtocol::handleReadData,this,session));
	return 0;
}

int csg::CCsgProtocol::handleSendData(const CSessionPtr session, const void* data, const int len)
{
	CAutoLock l(_writeLock);
	_sendBuffer->append(data, len);
	session->getSocket()->async_write_some(boost::asio::null_buffers(), boost::bind(&CCsgProtocol::handleWriteDataEx,this, session, boost::asio::placeholders::error));
	return 0;
}

int csg::CCsgProtocol::handleReadData(const CSessionPtr session)
{
	CAutoLock l(_readLock);
	do
	{
		if (_receiveHead)
		{
			if (_recvBuffer->getDataSize() < _protocolHead.msgSize)
			{
				//data not enough for a packet
				return true;
			}
			else
			{
				//data enough for a packet
				if (!handlePacket(session, _recvBuffer->getData() + SIZE_OF_PROTOCOL_HEAD, _protocolHead.msgSize))
				{
					LogInfo("CProtocol::handlePacket error");
					return false;
				}
				_recvBuffer->popData(_protocolHead.msgSize + SIZE_OF_PROTOCOL_HEAD);
				_receiveHead = false;
			}
		}
		else
		{
			if (_recvBuffer->getDataSize() >= SIZE_OF_PROTOCOL_HEAD)
			{
				memcpy(&_protocolHead, _recvBuffer->getData(), SIZE_OF_PROTOCOL_HEAD);
				_protocolHead.msgSize = endian(_protocolHead.msgSize);
				_receiveHead = true;
			}
			else
			{
				return true;
			}
		}
	} while (true);

	return false;

	return 0;
}

int csg::CCsgProtocol::handleWriteData(const CSessionPtr session)
{

	return 0;
}

int csg::CCsgProtocol::pushMessage(const CSessionPtr session,const CMsgBlockPtr& mb)
{
	CAutoSerializeStream  tmpOS(CSerializeStreamPool::instance()->newObject());

	_csg_write(*tmpOS, ERMIMessageTypeMQ);
	mb->_csg_write(*tmpOS);// _writeBody的时候写structType

	//整合发送前的flag数据
	tmpOS->prepareToAppend();

	CAutoSerializeStream  sendOs(CSerializeStreamPool::instance()->newObject());
	sendOs->writeSize(tmpOS->getFlagDataSize());
	sendOs->append(tmpOS->getFlagData(), tmpOS->getFlagDataSize());
	sendOs->append(tmpOS->getData(), tmpOS->getDataSize());

	//加协议头
	SProtocolHead head;
	head.msgSize = sendOs->getDataSize();
	CAutoSerializeStream addHeadOs(CSerializeStreamPool::instance()->newObject());
	addHeadOs->append(&head, SIZE_OF_PROTOCOL_HEAD);
	addHeadOs->append(sendOs->getData(), sendOs->getDataSize());

	return handleSendData(session, addHeadOs->getData(), addHeadOs->getDataSize());
}

int csg::CCsgProtocol::handlePacket(const CSessionPtr session, const void *packageData, const int len)
{
	CAutoSerializeStream is(CSerializeStreamPool::instance()->newObject());
	is->append(packageData, len);

	is->setUseBitMark(false);

	is->prepareToRead();

	ERMIMessageType mqType;
	_csg_read(*is, mqType);

	switch (mqType)
	{
	case csg::ERMIMessageTypeMQ:
	{
		CMsgBlockPtr msg = new CMsgBlock();
		msg->_csg_read(*is);

		if (!msg)
		{
			assert(false);
			return false;
		}
		msg->_msgBase->print();
	}
	break;
	case csg::ERMIMessageTypeCall:
	{
		SRMICall rmiCall;
		rmiCall._csg_read(*is);

		MapRMIObject mapObject;
		if (!CRMIObjectAdapter::instance()->findRmiObject("Test", mapObject))
		{
			return false;
		}
		MapRMIObject::iterator it = mapObject.find(rmiCall.rpcId);
		if (it == mapObject.end())
		{
			return false;
		}
		is->setUseBitMark(true);
		it->second->__onCall(session, rmiCall, *is);
	}
	break;
	case csg::ERMIMessageTypeCallRet:
	{
		SRMIReturn rmiReturn;
		rmiReturn._csg_read(*is);

		is->setUseBitMark(true);

		LogInfo("CProtocol::handlePacket" << ",messageId=" << rmiReturn.messageId);

		if (rmiReturn.messageId <= 0)
			return true;

		CRMIObjectBindPtr backObject;
		if (!session->getCallBackObject(rmiReturn.messageId, backObject))
		{
			return false;
		}
		if (backObject->_callBack)
		{
			backObject->_callBack->__response(*is);
		}
	}
	break;
	default:
	{
		LogErr("unkonw type=" << mqType);
		assert(false);
		return false;
	}
	}

	return true;
}

int csg::CCsgProtocol::handleWriteDataEx(const CSessionPtr session, boost::system::error_code err)
{
	if (err)
	{
		LogErr("CCsgProtocol::handleWriteDataEx handleWritet,ex=" << err.message());
		return -1;
	}
	else
	{
		LogDebug("CCsgProtocol::handleWriteDataEx handleWritet.....");
		boost::system::error_code write_err;
		int wlen = session->getSocket()->write_some(boost::asio::buffer(_sendBuffer->getData(),_sendBuffer->getDataSize()), write_err);
		if (write_err)
		{
			LogErr("CCsgProtocol::handleWriteDataEx write some2,ex=" << write_err.message());
			return -1;
		}
		else
		{
			LogDebug("CCsgProtocol::handleWriteDataEx handleWritet complete,size=" << wlen);
		}
		_sendBuffer->popData(wlen);
	}
	return 0;
}

File: 109/Client/assets/scripts/engine/net/protocol/CsgProtocol.js

import {CSerializeStream} from '../../serializeStream/CSerializeStream'
import {SProtocolHead,SIZE_OF_PROTOCOL_HEAD,ERMIMessageType,SRMIReturn,ERMIDispatchResult} from '../../rmi/CRmiDef'
import {CMsgManager} from '../../mq/CMsgManager'
import {CMsgBlock} from '../../mq/CMsgBlock'
import {CException,ECSGEErrorCode} from '../../exception/CException'
import {CRMIProxyCallBackObject} from '../../rmi/CRmiObject'

class CsgProtocol{
	_recvBuffer;
	_sendBuffer;
	_receiveHead;
	_protocolHead;
	constructor(){
		this._receiveHead=false;
		this._sendBuffer=new CSerializeStream();
		this._recvBuffer=new CSerializeStream();
		this._protocolHead=new SProtocolHead();
	}

	handleRecvData(session,data,len){
		this._recvBuffer.append(data,len);
		this.handleReadData(session);
	}

	handleReadData(session){
		do{
			if(this._receiveHead){
				if(this._recvBuffer.getDataSize()<this._protocolHead.msgSize+SIZE_OF_PROTOCOL_HEAD){
					return 0;
				}
				try{
					if(-1===this.handlePacket(session)){

						return -1;
					}
				}catch(err){
					console.log("CsgProtocol.handlePacket error,message="+err.message);
					throw err;
				}
				this._recvBuffer.popData(this._protocolHead.msgSize+SIZE_OF_PROTOCOL_HEAD);
				this._recvBuffer.reset();
				this._receiveHead=false;
			}else{
				if(this._recvBuffer.getDataSize()>=SIZE_OF_PROTOCOL_HEAD){
					this._recvBuffer.initReadView();
					this._protocolHead._csg_read(this._recvBuffer);
					this._receiveHead=true;
				}else{
					return 0;
				}
			}

		}while(true);
	}

	handlePacket(session){
		let is=new CSerializeStream();
		is.append(this._recvBuffer.getData(),this._protocolHead.msgSize+SIZE_OF_PROTOCOL_HEAD);
		//pop head data first
		is.popData(SIZE_OF_PROTOCOL_HEAD);
		is.initReadView();
		is.setUseBitMark(false);
		is.prepareToRead();

		let rmiType=is.readByte();
		console.log("rmi.type="+rmiType);
		switch(rmiType){
			case ERMIMessageType.ERMIMessageTypeMQ:
			{
				console.log("get mq from server");
				let msg=new CMsgBlock();
				msg._csg_read(is);

			}
			break;
			case ERMIMessageType.ERMIMessageTypeCall:
			{
				console.log("get call from server?");
				throw new CException(ECSGEErrorCode.ExceptionCodeRMIBase,"client shouldn't receive call type");
			}
			break;
			case ERMIMessageType.ERMIMessageTypeCallRet:
			{
				console.log("call back from server")
				let rmiReturn=new SRMIReturn();
				rmiReturn._csg_read(is);
				is.setUseBitMark(true);
				if(rmiReturn.messageId<=0)
					return 0;
				let backObject=session.getCallBackObject(rmiReturn.messageId);
				if(backObject=="undefined"){
					return -1;
				}
				if(backObject._callBack!="undefined"){
					if(ERMIDispatchResult.ERMIDispatchResultOk==rmiReturn.dispathStatus){
						backObject._callBack.__response(is);
					}else{
						backObject._callBack.__exception(is);
					}
				}
				
			}
			break;
			default:
			{
				console.log("unknow rmiType,type="+rmiType);
			}

		}


		return 0;
	}

}

export {CsgProtocol}

